File: src/metadata.rs
use crate::ast::{Distro, Variant};
use crate::error::Diagnostic;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PackageManager {
    Apk,
    Apt,
    Dnf,
    None,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BaseMetadata {
    pub mutable: bool,
    pub package_manager: PackageManager,
    pub shell_available: bool,
}

pub fn metadata_for(distro: Distro, variant: Variant) -> Result<BaseMetadata, Diagnostic> {
    if distro == Distro::Windows {
        return Err(Diagnostic::new(
            "windows base images are not supported by the Dockerfile backend yet",
            None,
        ));
    }

    let package_manager = match distro {
        Distro::Wolfi | Distro::Alpine => PackageManager::Apk,
        Distro::Debian | Distro::Ubuntu => PackageManager::Apt,
        Distro::Rhel => PackageManager::Dnf,
        Distro::Distroless | Distro::Windows => PackageManager::None,
    };
    let shell_available = !matches!(distro, Distro::Distroless | Distro::Windows);
    let mutable = match variant {
        Variant::Dev => true,
        Variant::Runtime => false,
    };

    Ok(BaseMetadata {
        mutable,
        package_manager,
        shell_available,
    })
}

pub fn package_for_tool(tool: &str, distro: Distro) -> Option<&'static str> {
    match distro {
        Distro::Wolfi | Distro::Alpine => match tool {
            "git" => Some("git"),
            "curl" => Some("curl"),
            "aws-cli" => Some("aws-cli"),
            "ca-certificates" => Some("ca-certificates"),
            "openssl" => Some("openssl"),
            "build-base" => Some("build-base"),
            "unzip" => Some("unzip"),
            _ => None,
        },
        Distro::Debian | Distro::Ubuntu => match tool {
            "git" => Some("git"),
            "curl" => Some("curl"),
            "aws-cli" => Some("awscli"),
            "ca-certificates" => Some("ca-certificates"),
            "openssl" => Some("openssl"),
            "build-base" => Some("build-essential"),
            "unzip" => Some("unzip"),
            _ => None,
        },
        Distro::Rhel => match tool {
            "git" => Some("git"),
            "curl" => Some("curl"),
            "aws-cli" => Some("awscli"),
            "ca-certificates" => Some("ca-certificates"),
            "openssl" => Some("openssl"),
            "build-base" => Some("@development-tools"),
            "unzip" => Some("unzip"),
            _ => None,
        },
        Distro::Distroless | Distro::Windows => None,
    }
}

pub fn carried_tool_paths(tool: &str) -> Option<&'static [&'static str]> {
    match tool {
        "git" => Some(&["/usr/bin/git", "/usr/libexec/git-core"]),
        "ca-certificates" => Some(&["/etc/ssl/certs"]),
        "aws-cli" => Some(&["/usr/bin/aws", "/usr/lib/aws-cli"]),
        _ => None,
    }
}

File: ddsl-core/src/main/java/ie/bitstep/ddsl/metadata/Metadata.java
package ie.bitstep.ddsl.metadata;

import ie.bitstep.ddsl.ast.Ast;
import ie.bitstep.ddsl.diagnostics.DdslException;
import ie.bitstep.ddsl.model.Model;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class Metadata {
  private static final Map<String, List<String>> CARRIED_TOOL_PATHS = Map.of(
      "git", List.of("/usr/bin/git", "/usr/libexec/git-core"),
      "ca-certificates", List.of("/etc/ssl/certs"),
      "aws-cli", List.of("/usr/bin/aws", "/usr/lib/aws-cli"));

  private Metadata() {}

  public static Model.BaseMetadata forBase(Ast.Distro distro, Ast.Variant variant) {
    if (distro == Ast.Distro.WINDOWS) {
      throw new DdslException("windows base images are not supported by the Dockerfile backend yet", null);
    }
    Model.PackageManager packageManager = switch (distro) {
      case WOLFI, ALPINE -> Model.PackageManager.APK;
      case DEBIAN, UBUNTU -> Model.PackageManager.APT;
      case RHEL -> Model.PackageManager.DNF;
      case DISTROLESS, WINDOWS -> Model.PackageManager.NONE;
    };
    boolean shellAvailable = distro != Ast.Distro.DISTROLESS && distro != Ast.Distro.WINDOWS;
    boolean mutable = variant == Ast.Variant.DEV;
    return new Model.BaseMetadata(mutable, packageManager, shellAvailable);
  }

  public static Optional<String> packageForTool(String tool, Ast.Distro distro) {
    return switch (distro) {
      case WOLFI, ALPINE -> switch (tool) {
        case "git", "curl", "aws-cli", "ca-certificates", "openssl", "build-base", "unzip" -> Optional.of(tool);
        default -> Optional.empty();
      };
      case DEBIAN, UBUNTU -> switch (tool) {
        case "git", "curl", "ca-certificates", "openssl", "unzip" -> Optional.of(tool);
        case "aws-cli" -> Optional.of("awscli");
        case "build-base" -> Optional.of("build-essential");
        default -> Optional.empty();
      };
      case RHEL -> switch (tool) {
        case "git", "curl", "ca-certificates", "openssl", "unzip" -> Optional.of(tool);
        case "aws-cli" -> Optional.of("awscli");
        case "build-base" -> Optional.of("@development-tools");
        default -> Optional.empty();
      };
      case DISTROLESS, WINDOWS -> Optional.empty();
    };
  }

  public static Optional<List<String>> carriedToolPaths(String tool) {
    return Optional.ofNullable(CARRIED_TOOL_PATHS.get(tool));
  }
}
